cmd/node_agent/na/gcp_test.go

```diff
@@ -27,28 +28,93 @@ const (
+	errorMessage string
 }
 
 // A mock fetcher for FetchToken.
 func (fetcher *mockTokenFetcher) FetchToken() (string, error) {
-	return token, nil
+	var err error
+
+	if len(fetcher.errorMessage) > 0 {
+		err = fmt.Errorf(fetcher.errorMessage)
+	}
+
+	return token, err
 }
 
 func TestGetDialOptions(t *testing.T) {
-	gcp := gcpPlatformImpl{&mockTokenFetcher{}}
-	config := &Config{RootCACertFile: "testdata/cert-chain-good.pem"}
-	options, err := gcp.GetDialOptions(config)
-	if err != nil {
-		t.Error(err)
+	testCases := map[string]struct {
+		config         *Config
+		optionsLen     int
+		token          string
+		tokenFetchErr  string
+		expectedErr    string
+		expectedOption uintptr
+	}{
+		"nil configuration": {
+			config: &Config{
+				RootCACertFile: "testdata/cert-chain-good.pem",
+			},
+			optionsLen:    0,
+			token:         "abcdef",
+			expectedErr:   "Nil configuration passed",
+			tokenFetchErr: "Nil configuration passed",
+		},
+		"Token fetch error": {
+			config: &Config{
+				RootCACertFile: "testdata/cert-chain-good.pem",
+			},
+			optionsLen:    0,
+			token:         "",
+			expectedErr:   "Nil configuration passed",
+			tokenFetchErr: "Nil configuration passed",
+		},
+		"Root certificate file read error": {
+			config: &Config{
+				RootCACertFile: "testdata/cert-chain-good_not_exist.pem",
+			},
+			optionsLen:    0,
+			token:         token,
+			tokenFetchErr: "",
+			expectedErr:   "open testdata/cert-chain-good_not_exist.pem: no such file or directory",
+		},
+		"Token fetched": {
+			config: &Config{
+				RootCACertFile: "testdata/cert-chain-good.pem",
+			},
+			optionsLen:     2,
+			token:          token,
+			tokenFetchErr:  "",
+			expectedOption: reflect.ValueOf(grpc.WithPerRPCCredentials(&jwtAccess{token})).Pointer(),
+		},
 	}
 
-	// Make sure there're two dial options, one for TLS and one for JWT.
-	if len(options) != 2 {
-		t.Errorf("Wrong dial options size")
-	}
+	for id, c := range testCases {
+		gcp := gcpPlatformImpl{&mockTokenFetcher{c.token, c.tokenFetchErr}}
+
+		options, err := gcp.GetDialOptions(c.config)
+		if len(c.expectedErr) > 0 {
+			if err == nil {
+				t.Errorf("Succeeded. Error expected: %v", err)
+			} else if err.Error() != c.expectedErr {
+				t.Errorf("%s: incorrect error message: %s VS %s",
+					id, err.Error(), c.expectedErr)
+			}
+		} else if err != nil {
+			t.Fatalf("Unexpected Error: %v", err)
+		}
+
+		// Make sure there're two dial options, one for TLS and one for JWT.
+		if len(options) != c.optionsLen {
+			t.Errorf("Wrong dial options size. Exptcted %v, Actual %v", c.optionsLen, len(options))
+		}
 
-	expectedOption := grpc.WithPerRPCCredentials(&jwtAccess{token})
-	if reflect.ValueOf(options[0]).Pointer() != reflect.ValueOf(expectedOption).Pointer() {
-		t.Errorf("Wrong option found")
+		if len(options) > 0 {
+			expectedOption := grpc.WithPerRPCCredentials(&jwtAccess{token})
+			if reflect.ValueOf(options[0]).Pointer() != reflect.ValueOf(expectedOption).Pointer() {
```

Comment:
It seems you are only testing options[0].
Can you make expectedOption []uintptr, so that you can test both. Also you can remove optionsLen.

Reply:
Applied